File: Receiver/src/Communication/BluetoothManager.cpp
/******************************************************************************
 * Proyecto : MarioKart ESP32 RC
 * Archivo  : BluetoothManager.cpp
 * Autor    : Narciso Ivan Cisneros Acosta
 *
 * Descripción:
 * Implementación del administrador del protocolo Bluetooth.
 ******************************************************************************/

#include "BluetoothManager.h"

namespace MK
{

//=============================================================================
// Ciclo de vida
//=============================================================================

bool BluetoothManager::Begin(
    const char* deviceName) noexcept
{
    Serial.printf("BTM Begin.");
    return m_transport.Begin(
        deviceName);
}

//=============================================================================
// Estado
//=============================================================================

bool BluetoothManager::Connected() const noexcept
{
    Serial.printf("BTM Connected.");
    return m_transport.Connected();
}

//=============================================================================
// DriverCommand
//=============================================================================

bool BluetoothManager::Receive(
    Protocol::DriverCommand& command) noexcept
{
    Serial.println("1");

    // std::uint8_t buffer[
    //     Protocol::PacketSize<
    //         Protocol::DriverCommand>()];

    std::uint8_t buffer[7];            

    Serial.println("2");

    // const std::size_t received =
    //     m_transport.Receive(
    //         buffer,
    //         sizeof(buffer));

    const std::size_t received =
    m_transport.Receive(
        buffer,
        7);

    Serial.printf(
        "Received=%u Expected=%u\n",
        received,
        sizeof(buffer));

    //Serial.printf("RX Type=%u  Expected=%u\n",(uint8_t)packet.header.type,(uint8_t)Protocol::PacketType::DriverCommand);

    Serial.printf("sizeof(PacketHeader)=%u\n", sizeof(Protocol::PacketHeader));
    Serial.printf("sizeof(DriverCommand)=%u\n", sizeof(Protocol::DriverCommand));
    Serial.printf("PacketSize=%u\n", Protocol::PacketSize<Protocol::DriverCommand>());

    if(received > 0)
    {
        Serial.print("RX = ");

        for(std::size_t i=0;i<received;i++)
        {
            Serial.printf("%02X ",buffer[i]);
        }

        Serial.println();
    }

    if(received != sizeof(buffer))
    {
        Serial.println("3");

        return false;
    }

    Serial.println("4");

    Protocol::Packet<
        Protocol::DriverCommand> packet;

    if(!Protocol::PacketSerializer::Deserialize(
            buffer,
            received,
            packet))
    {
        Serial.println("5");

        return false;
    }

    Serial.println("6");

    if(packet.header.type !=
       Protocol::PacketType::DriverCommand)
    {
        Serial.printf(
            "Wrong Type=%u\n",
            (uint8_t)packet.header.type);

        return false;
    }

    Serial.println("7");

    command =
        packet.payload;

    return true;
}

//=============================================================================
// VehicleStatus
//=============================================================================

bool BluetoothManager::Send(
    const Protocol::VehicleStatus& status) noexcept
{
    Serial.printf("BTM Send Begin.");
    //-------------------------------------------------------------
    // Construir Packet
    //-------------------------------------------------------------

    Protocol::Packet<
        Protocol::VehicleStatus> packet;

    packet.header.type =
        Protocol::PacketType::VehicleStatus;

    packet.header.payloadSize =
        sizeof(Protocol::VehicleStatus);

    packet.payload = status;

    //-------------------------------------------------------------
    // Serializar
    //-------------------------------------------------------------

    std::uint8_t buffer[
        Protocol::PacketSize<
            Protocol::VehicleStatus>()];

    if (!Protocol::PacketSerializer::Serialize(
            packet,
            buffer,
            sizeof(buffer)))
    {
        return false;
    }

    //-------------------------------------------------------------
    // Enviar
    //-------------------------------------------------------------
    Serial.printf("BTM Send End.");
    return m_transport.Send(
        buffer,
        sizeof(buffer));
}

}

File: AndroidController/MarioKart.Android/MarioKart.Android/Communication/Bluetooth/BluetoothManager.cs
/******************************************************************************
 * Proyecto : MarioKart ESP32 RC
 * Archivo  : BluetoothManager.cs
 * Autor    : Narciso Ivan Cisneros Acosta
 *
 * Descripción:
 * Gestiona el protocolo Bluetooth.
 *
 * No conoce BluetoothSocket ni Streams.
 * Únicamente trabaja con paquetes del protocolo.
 ******************************************************************************/

using System;
using System.Threading;
using System.Threading.Tasks;

using MarioKart.Android.Communication.Protocol;
using MarioKart.Android.Debug;

namespace MarioKart.Android.Communication.Bluetooth
{
    public sealed class BluetoothManager
    {
        //=====================================================================
        // Campos
        //=====================================================================

        private readonly BluetoothTransport m_transport;

        private CancellationTokenSource m_receiveLoopCancellation;

        //=====================================================================
        // Eventos
        //=====================================================================

        public event EventHandler<Packet> PacketReceived;

        //=====================================================================
        // Constructor
        //=====================================================================

        public BluetoothManager(
            BluetoothTransport transport)
        {
            m_transport =
                transport
                ?? throw new ArgumentNullException(
                    nameof(transport));
        }

        //=====================================================================
        // Estado
        //=====================================================================

        public bool IsConnected
        {
            get
            {
                return m_transport.IsConnected;
            }
        }

        //=====================================================================
        // Recepción
        //=====================================================================

        public void Start()
        {
            if (m_receiveLoopCancellation != null)
            {
                return;
            }

            m_receiveLoopCancellation =
                new CancellationTokenSource();

            _ = ReceiveLoopAsync(
                m_receiveLoopCancellation.Token);
        }

        public void Stop()
        {
            m_receiveLoopCancellation?.Cancel();

            m_receiveLoopCancellation = null;
        }

        //=====================================================================
        // Envío
        //=====================================================================

        public async Task SendAsync(
            Packet packet)
        {
            if (packet == null)
            {
                return;
            }

            byte[] buffer =
                PacketSerializer.Serialize(packet);

            if (buffer == null)
            {
                return;
            }

            await m_transport.SendAsync(
                buffer);
        }
        //=====================================================================
        // Desconexión
        //=====================================================================

        public void Disconnect()
        {
            //-------------------------------------------------------------
            // Detener recepción
            //-------------------------------------------------------------

            Stop();

            //-------------------------------------------------------------
            // Cerrar transporte
            //-------------------------------------------------------------

            m_transport.Disconnect();
        }

        //=====================================================================
        // Receive Loop
        //=====================================================================

        private async Task ReceiveLoopAsync(
            CancellationToken cancellationToken)
        {
            byte[] buffer =
                new byte[256];

            while (!cancellationToken.IsCancellationRequested)
            {
                int count =
                    await m_transport.ReceiveAsync(
                        buffer);

                if (count <= 0)
                {
                    continue;
                }

                ProcessReceivedData(
                    buffer,
                    count);
            }
        }

        //=====================================================================
        // Procesamiento
        //=====================================================================

        private void ProcessReceivedData(
            byte[] buffer,
            int length)
        {
            Packet packet =
                PacketSerializer.Deserialize(
                    buffer.AsSpan(0, length));

            if (packet == null)
            {
                return;
            }
            PacketReceived?.Invoke(
                this,
                packet);
        }
    }
}